python-package/lightgbm/callback.py

```diff
@@ -291,32 +296,49 @@ def _gt_delta(self, curr_score: float, best_score: float, delta: float) -> bool:
     def _lt_delta(self, curr_score: float, best_score: float, delta: float) -> bool:
         return curr_score < best_score - delta
 
-    def _is_train_set(self, ds_name: str, eval_name: str, train_name: str) -> bool:
-        return (ds_name == "cv_agg" and eval_name == "train") or ds_name == train_name
+    def _is_train_set(self, ds_name: str, eval_name: str, env: CallbackEnv) -> bool:
+        """Check, by name, if a given Dataset is the training data."""
+        # for lgb.cv() with eval_train_metric=True, evaluation is also done on the training set
+        # and those metrics are considered for early stopping
+        if ds_name == "cv_agg" and eval_name == "train":
+            return True
+
+        # for lgb.train(), it's possible to pass the training data via valid_sets with any eval_name
+        if isinstance(env.model, Booster):
+            if ds_name == env.model._train_data_name:
+                return True
+
+        return False
 
     def _init(self, env: CallbackEnv) -> None:
         if env.evaluation_result_list is None or env.evaluation_result_list == []:
             raise ValueError(
                 "For early stopping, at least one dataset and eval metric is required for evaluation"
             )
+
+        if self.stopping_rounds <= 0:
+            raise ValueError(f"stopping_rounds should be greater than zero. got: {self.stopping_rounds}")
```

Comment:
I agree! Having it be a loud error right when the callback is created, instead of deferred all the way til the first iteration of training, seems useful. And I'd be surprised to learn that there are other libraries or user code depending on initializing `lgb.early_stopping()` with a negative value of this and then somehow updating the value before the first time it's called.

Moved into `__init__()` in https://github.com/microsoft/LightGBM/pull/6114/commits/bd3366a6fa9335e3d935e6b91fd6d8db25dea304.

Reply:
Doing this broke this test:

https://github.com/microsoft/LightGBM/blob/f175cebd2d971ca6140455035e526d67f04237eb/tests/python_package_test/test_engine.py#L4502-L4507

Now the error from the early stopping callback gets thrown before this one from the C++ side:

https://github.com/microsoft/LightGBM/blob/f175cebd2d971ca6140455035e526d67f04237eb/src/io/config.cpp#L37

So I pushed https://github.com/microsoft/LightGBM/pull/6114/commits/7a98d82a31c94cdbfcbd2a8cd0c4a557281305a3, which:

* switches that test to use a different parameter, to keep covering that C++-side validation
* adds a test in `test_callback.py` on this specific error from `lgb.early_stopping()`
* adds an `isinstance()` check in the condition guarding that error in `lgb.early_stopping()`, so you can an informative error instead of something like `TypeError: '<=' not supported between instances of 'str' and 'int'`

Given all those changes, @jmoralez could you re-review? I don't want to sneak those in on your previous approval.